compiler/next/include/chpl/types/Type.h

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_TYPES_TYPE_H
+#define CHPL_TYPES_TYPE_H
+
+#include "chpl/queries/Context.h"
+#include "chpl/types/TypeClasses.h"
+#include "chpl/types/TypeTag.h"
+
+namespace chpl {
+namespace types {
+
+
+/**
+  This is the base class for classes that represent a type.
+
+  Functions like someType->isRecord() / someAst->toRecord() are available and
+  generated for all Typnode subclasses.
+
+ */
+class Type {
```

Comment:
Could we introduce a parent class `e.g.` `QueryObject` which requires the `contentsMatchInner` and `markUniqueStringsInner` methods? It seems like there's more than one class of object now which requires its memory to be managed by the query framework.

Reply:
I think that's theoretically possible but I'm not sure of the value of it and I'd like to put it off for the moment.